linked_list/LinkedList.c

```diff
@@ -0,0 +1,249 @@
+struct Node* deleteAtPos(struct Node **start, int pos) {
+	if (*start == NULL) {
+		return NULL;
+	} else {
+		struct Node* temp = *start;		// temp node to go through each node
+		int i = 1;
+		while (temp->next->next != NULL && i < pos-1) {	// until we reach to the position or last element of list is encountered
+			temp = temp->next;
+			i++;
+		}
+		if (i != pos-1) {
+			printf("Invalid Index\n\n");
+			return NULL;
+			} else {
+			struct Node* tmp = temp->next;
+		    	temp->next = temp->next->next;	// Replacing address of position 'pos' with  node next to 'pos'
+			return tmp;
+		}
+	}
+}
+
+//  To traverse whole list
+void traverse(struct Node *start) {
+    printf("Linked List\n");
+	if(start == NULL) {
+		printf("Linked List is Empty");
+	} else {
+		struct Node* temp = start;		// temp node to go through each node
+		while (temp->next != NULL) {
+			printf(" %d ->", temp->d);
+			temp = temp->next;
+		}
+		printf("%d", temp->d);
+	}
+	printf("\n\n");
+}
+
+//  To search an element
+int search(struct Node *start, int data) {
+	struct Node* temp = start;
+	int found = 0, i = 1;
+	while (temp->next != NULL) {
+		if(temp->d == data) {
+			found = 1;
+			break;
+		}
+		i++;
+		temp = temp->next;
+	}
+	if (found != 0) {
+		return i;
+	} else {
+		return -1;
+	}
+}
+
+// to print search result
+void searchResult(int d, int pos) {
+	if(pos == -1) { 
+		printf("Element %d is not in list\n", d);
+	} else {
+		printf("Element %d is found at position %d\n", d, pos);
+	}
+	printf("\n");
+}
+
+void deleteResult(struct Node** node) {
+	if(*node == NULL) {
+		printf("Deletion is unsuccessfull\n");
+	} else {
+		printf("Node %d is deleted\n", (*node)->d);
+		free(*node);
+	}
+	printf("\n");
+}
+
+void insertResult(int res, int d) {
+	if(res == 0) {
+		printf("Invalid Index\n");
+	} else {
+		printf("Element %d is inserted\n", d);
+	}
+	printf("\n");
+}
+
+int main() {
+	struct Node *start = NULL;
+	struct Node *node;
+	int res;
+	// Adding element to the list
+	if (getNewNode(23) != NULL) {
+		res = addFront(&start, getNewNode(23));
+		insertResult(res, 12);
+	}
+	traverse(start);
+	
+	// Adding elements at the end
+	if (getNewNode(25) != NULL) {
+		res = addLast(&start, getNewNode(25));
+		insertResult(res, 25);
+	}
+	if (getNewNode(46) != NULL) {
+		res = addLast(&start, getNewNode(46));
+		insertResult(res, 46);
+	}
+	traverse(start);
+	
+	// Adding elements to front
+	if (getNewNode(12) != NULL) {
+		res = addFront(&start, getNewNode(12));
+		insertResult(res, 12);	
+	}
+	if (getNewNode(11) != NULL) {
+		res = addFront(&start, getNewNode(11));
+		insertResult(res, 11);	
+	}
+	traverse(start);
+
+	// Adding element at a given index (Adding 14 at index 3)
+	if (getNewNode(14) != NULL) {
+		res = addAtPos(&start, getNewNode(14), 3);
+		insertResult(res, 14);	
+	}
+	traverse(start);
+
+	// Removing First Element
+	node = deleteFirst(&start);
+	deleteResult(&node);
+	traverse(start);
+
+	// search
+	res = search(start, 14);
+	searchResult(14, res);
```

Comment:
Why can't we have a single function here which returns node pointer if found else null.